pipeline/objects/decorators.py

```diff
@@ -44,26 +47,49 @@ def execute_func(*args, **kwargs):
                 elif hasattr(input_arg, "__pipeline_model__"):
                     if function.__pipeline_function__.class_instance is None:
                         function.__pipeline_function__.class_instance = input_arg
+                elif isinstance(input_arg, tuple) and all(
+                    isinstance(var, Variable) for var in input_arg
+                ):
+                    raise Exception(
+                        "Must seperate outputs from functions with Tuple outputs:"
+                        "`var1, var2, ..., varN = func(...)`",
+                    )
+
                 else:
                     raise Exception(
                         (
-                            "You can't input random variables, "
-                            "follow the way of the Pipeline. Got type %s"
-                            % type(input_arg)
+                            f"Can only input pipeline variables to a function"
+                            f"when defining a graph, got: {type(input_arg)}"
                         )
                     )
 
-            node_output = Variable(type_class=function.__annotations__["return"])
-            Pipeline.add_variable(node_output)
+            node_outputs: typing.List[Variable] = []
+
+            function_output = function.__annotations__["return"]
+            if getattr(function_output, "__origin__", None) == tuple:
+                context_manager_variables = node_outputs = tuple(
+                    Variable(type_class=output_variable)
+                    for output_variable in function_output.__args__
+                )
+
+            else:
+                context_manager_variables = Variable(
+                    type_class=function.__annotations__["return"]
+                )
+                node_outputs = [context_manager_variables]
+
+            Pipeline.add_variables(*node_outputs)
             Pipeline.add_function(function.__pipeline_function__)
+
             new_node = GraphNode(
                 function=function.__pipeline_function__,
                 inputs=processed_args,
-                outputs=[node_output],
+                outputs=[*node_outputs],
```

Comment:
As `node_outputs` is already an iterable, is `outputs=node_outputs` sufficient?

Reply:
Nice catch, this was some lingering code left partway through adding this in.
I added an extra test to check for this working as well.